Design Doc Review for Small Batch Publishing (#25956)
department-of-veterans-affairs/va.gov-team: platform/engineering/design-docs/2021-06-11-small-batch-publishing.md

* [Specifics](#specifics)
  * [Detailed Design](#detailed-design)
    * [Drupal Triggers Content Publishing](#drupal-triggers-content-publishing)
    * [Preview Server Changes](#preview-server-changes)
    * [Facility Banner Alerts](#facility-banner-alerts)
    * [Metalsmith plugins](#metalsmith-plugins)
    * [Publish Batch API](#publish-batch-api)
    * [Continuous Integration](#continuous-integration)
    * [Assets](assets)
  * [Code Location](code-location)
  * [Testing Plan](#testing-plan)
  * [Logging](#logging)
  * [Debugging](#debugging)
  * [Caveats](#caveats)
  * [Security Concerns](#security-concerns)
  * [Privacy Concerns](#privacy-concerns)
  * [Open Questions and Risks](#open-questions-and-risks)
  * [Work Estimates](#work-estimates)
  * [Alternatives, Research, and Future Work](#alternatives-Research-and-future-work)
* [Long Term Vision](#long-term-vision)
* [Revision History](#revision-history)

## **Overview**

### **Objective**

An editor on the CMS should be able to publish content to va.gov when the "Publish" button is pressed in the CMS. This document outlines an approach utilizing the preview server’s code to render HTML which is directly deployed to S3.

> This design doc only covers publishing pages which are represented in Drupal as a node. The full content build still runs hourly.

Reviewer: ```suggestion
This design doc only covers publishing pages which are represented in Drupal as a node. The full content-only build still runs 9am-12pm Hourly, 1:45pm, 4pm, 5pm ET M-F.
```

Reviewer: What does the content-only deploy contain if all the drupal content goes out immediately? The vagov-content stuff? If that's the case, I feel like we can just cancel the content-only deploys, that repo is barely ever updated

Author: We are planning on rolling out SBP by content type over a period of months.  After that, there are the pages from the markdown content repo and finally, pages (like listing and location pages) that use Drupal content to build a page but are not represented in Drupal.  Though https://github.com/department-of-veterans-affairs/va.gov-team/issues/25940 from @melissajschaff  might bring a solution.  Either a dependency graph in the build or moving the pages to Drupal would help with this use case.

Reviewer: I think we should try and push for more unification of the content so we actually turn off the "hourly" content release. Seems like there would be a ton of times throughout the day were a build were happening that would block instant publishing making it a little more annoying rather than helpful

Author: I updated the plan to cutover content types from full build to SPB. This will make reduce lots of complexity.

The primary editor use case is for updating currently published content.

#### Links

* [Initiative in Zenhub](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5499) 
* [Accelerated Publishing Initiave](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/4746)
* [Accelerated Publishing Discovery](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/4748)
* [Single Page Preview Discovery](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5095)
* [Accelerated Publishing Documentation](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/)
* [Accelerated Publhsing Discovery Finding](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/cms/accelerated_publishing/findings_recommendations.md)

### **Background**

> The current content-build for va.gov runs once an hour and generates all HTML pages. The current build takes approximatly 15 minutes for page to show up on va.gov. When the amount of content in Drupal reaches 48K nodes, the build will take close to 45 minutes to publish content to va.gov.  Editors are often confused about when the content they published in the CMS will be live on va.gov.

Reviewer: ```suggestion
The current content-only build for va.gov runs once an hour and generates all HTML pages. The current build takes approximately 15 minutes for page to show up on va.gov. When the amount of content in Drupal reaches 48K nodes, the build will take close to 1 hour to publish content to va.gov.  Editors are often confused about when the content they published in the CMS will be live on va.gov.
```

> Editors coming from Team Sites expect publication of content within 5 minutes.  As editors migrate from Team Sits to Drupal, editors often do not understand why the content they publish does not show up on va.gov in a timely mannter.

Reviewer: ```suggestion
Editors coming from Team Sites expect publication of content within 5 minutes.  As editors migrate from Team Sites to Drupal, editors often do not understand why the content they publish does not show up on va.gov in a timely mannter.
```

### **High Level Design**

The same code used to generate the HTML for the preview server can be used to generate the HTML for va.gov.  

The following diagram shows a high level design of how a node in Drupal would be published “immediately” on va.gov.

> Drupal is in charge of determining when and which pages need to be published based upon the action by an editor.  When Drupal determines a page is to be published, a Publish Batch Record will be sent to Amazon Simple Queue Service (SQS). A Lambda job will pick up the Publish Batch Record to build the HTML. After the HTML has been created, it’s sent to an S3 bucket which serves the html for va.gov.  

Reviewer: ```suggestion
Drupal is in charge of determining when and which pages need to be published based upon the action by an editor.  When Drupal determines a page is to be published, a Publish Batch Record will be sent to Amazon Simple Queue Service (SQS). A Lambda job will pick up the Publish Batch Record to build the HTML. After the HTML has been created, it’s sent to an [S3 bucket](http://content.www.va.gov.s3-website-us-gov-west-1.amazonaws.com/) which serves the html for va.gov.  
```

The code used by the lambda to generate the HTML will utilize the current preview server code.  As part of this work, the preview server code will be abstracted and improved to provide production ready HTML.

![image](https://user-images.githubusercontent.com/121603/121699098-a2cd3b00-ca9c-11eb-8a0e-5e15cce7ce4d.png)

Editable link: [https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#](https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#)

## **Specifics**

### **Detailed Design**

#### Drupal Triggers Content Publishing

> When an editor performs an action in Drupal (ie publishing a node, edits a menu item), Drupal will determine which pages need to be published. [The logic will be encapsulated by a Drupal Service.](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5509) The rules around the dependency graph are being tracked on [here](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/dependency_graph) and are currently a work in progress. Which rules are addressed will be determined by the content type being worked on to support Small Batch Publishing.  Drupal will also have logic to not allow a Small Batch Publishing if a full build is in progress.

Reviewer: > Drupal will also have logic to not allow a Small Batch Publishing if a full build is in progress.

I'm curious about this implementation and the user affects. If we ever push that 1 hour build time the assumption is that not content will be published in that time period or maybe queued? and released after?

Author: This is a great question and something we have been discussing without a great solution.  

 Initially, the lambda will requeue the job if there is a build in-progress.  It should also set the Publish Batch Records `state` to `ON HOLD` or something.  Then when the lambda job runs, it can look at the revision that is live and compare it to the revision it's publishing to not publish an older revision.

Longer-term, hopefully, we can find a way to remove content types from the full build and only do them in the SBP.  Actually, thinking about this now, it's probably easier to do short-term.  To just cut over from full build to SBP.  @timwright12 thoughts about this?

Reviewer: For content, yeah, I'd love to just move it into drupal so metalsmith doesn't really even need to worry and can just focus on templates

Author: I updated the plan to cutover content types from full build to SPB.  This will make reduce lots of complexity.

Drupal will track the queue progress using the [advancedqueue](https://www.drupal.org/project/advancedqueue) and the Public Batch API described below. The advancedqueue module allows for observability and reporting.

> #### Preview Server Changes

Reviewer: Just food for thought that I think engineers use the preview functionality locally as well, so another point of QA

Author: I added notes about validating local build.

*   Code will be updated to use the json processing from here: `src/site/stages/build/drupal/*.js`
*   `/diff` endpoint code will be refactored to use the same code as the lambda function.
*   GraphQL Query building will be refactored to have Full and Small Batch use the same queries.
*   The Get Drupal Content plugin should be broken apart to make the layers reusable with the different services.

#### Facility Banner Alerts

The first type of data to be updated using the Small Batch Publishing will be the facility banner alerts and alerts. These types of content currently trigger a deploy from Drupal to the jenkins job. The trigger from Drupal to Jenkins will be replaced with Small Batch Publishing.  By removing the need for Drupal to trigger the Jenkins job, we eliminate logic around Drupal bulk publish events and when an editor is allowed to trigger a Small Batch Publish.

1. Banner Alert or Alert updated in Drupal
2. Drupal sends Publish Batch Record to SQS
3. Lambda picks up Publish Batch record
4. Metalsmith build runs for alerts and generates a JSON file instead of an HTML file.
5. JSON file is pushed to S3

> 6. A React Widget looks for the JSON file on S3 to populate the alert.

Reviewer: If this is a react widget on the page do we need metalsmith to run at all? Can we just update an endpoint and hit it live?

Author: The idea would that the react widget would load the data directly from a JSON file on S3 so no metasmith build required.  The JSON file would be updated as part of the SBP pipeline.

Reviewer: Gotcha, ok, #3 confused me there

Author: I separated out the creation of the react widget and linked to the epic with more detail.

| Add content assets                                        | X              |                            |                         |                                                                                                                                                         |
| Plug the content into the templates                       | X              | X                          | X                       |                                                                                                                                                         |
| Translate the markdown to html                            | X              | X                          |                         | This probably isn’t needed on SBP since the markdown pages will not be processed.                                                                       |
| Add permalinks and change foo.md to foo/index.html        | X              | X                          |                         |                                                                                                                                                         |

> | Create header and footer                                  | X              |                            |                         |                                                                                                                                                         |

Reviewer: There's an open ticket to add this to the preview functionality

Author: Would this requre the https://github.com/department-of-veterans-affairs/vagov-content repo?

Reviewer: TBD on that I think @theodur is picking up that ticket next week

Author: Is this the issue? https://github.com/department-of-veterans-affairs/va.gov-team/issues/25793

Reviewer: @indytechcook yeah that's the one

Author: Thanks, I linked it to the doc!

Check the [list of function aliases](https://docs.aws.amazon.com/lambda/latest/dg/API_ListAliases.html) for an alias matching the checksum. If an alias doesn't exist yet for the checksum:

*   [Update and publish the function code](https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionCode.html) using the archived function from S3 for dev and staging lambda functions.
*   [Create the alias](https://docs.aws.amazon.com/lambda/latest/dg/API_CreateAlias.html) for the new version of the function.

If the current branch is `master`, also get the version that the `master` alias points to. If the master alias doesn't match the version of the checksum alias:

*   [Update the master alias](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/lambda/update-alias.html) to the version currently aliased with the checksum.

A second Github Action will be created that is [manually triggered](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/). The action will [Update and publish the function code](https://docs.aws.amazon.com/lambda/latest/dg/API_UpdateFunctionCode.html) the production lambda function.

#### Assets

> Assets (images, pdfs) which are generated in Drupal will be uploaded to S3.  The content build will be updated to use the new S3 location as the `src` for the assets.  The details of this implementation will be in a different design doc and implemented outside of Small Batch Publishing.

Reviewer: > The content build will be updated to use the new S3 location as the `src` for the assets.

Ideally we'd hope that the `src` are already set to S3 before the content build starts so we don't have to parse it. Is that possible?

Author: Yeah.  Initially, we were thinking of adding a new field to the Drupal file entities which contained the full S3 path.  Then we update the build/templates to use the new field.  This would remove any processing in the build.  We discussed this idea with @alexandec earlier this week.

Author: I added more detail here and linked to the epic.

### **Code Location**

The Lambda function and its supporting code will live in the `content-build` repository.   The POC code will be updated to support dev/stage/prod.

The handler source code, bundling, and upload code from the POC lives in the branch `nh_preview_server_lambda` on this [PR](https://github.com/department-of-veterans-affairs/content-build/pull/177).

*   Lambda function: `src/platform/lambdas/single-page-build.js`
*   Webpack: `config/webpack.single-page-build.js`
*   GitHub Actions: `.github/workflows/single-page-build.yml`

The Lambda functions will be provisioned in the devops repo.  The POC code will be updated to support dev/stage/prod.

The Drupal code will live in the `va.gov-cms` repository.

### **Testing Plan**

> SBP and full build need to be tested and the output validated per content type.  The preview server currently has a `/diff` endpoint which generates a visual difference between the preview server HTML and production.  Example: [http://preview-dev.vfs.va.gov/diff?nodeId=2578](http://preview-dev.vfs.va.gov/diff?nodeId=2578). The preview will manually check for a subset of each content type before we turn on instant publishing for the content type.

Reviewer: > SBP and full build need to be tested 
Also local workflows

### **Logging**

GitHub Actions will output logs for the build and versioning of the function. Log output from the function (from calling console methods) will be stored in CloudWatch.

The lambda function will also communicate back to Druapl using the [Publish Batch API](#publish-batch-api) where Drupal will generate an Audit Trail of what was build by who and when.

### **Debugging**

*   One can start diagnosing any issues using the logs mentioned above.
*   The Webpack build of the function can be run locally to debug issues with the build.
*   The code should be abstracted in a way where the code can be executed outside of the lambda function.  For example, the preview server itself should be running the same code.

> *   Debugging the HTML produced and using the `/diff` endpoint which shows the HTML from the code compared to production.

Reviewer: We use timestamps on assets as a cache buster, so you likely won't be able to get a true diff comparison here: https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/stages/build/plugins/modify-dom/process-entry-names.js

Author: Yeah, those and the ids added by the analytics middleware.  When we manually validate, it's easy to see them on the diff.  We don't have an automated method in mind yet.

Thanks for the heads up.

| Check for CMS URLs                                        | X              |                            |                         |                                                                                                                                                         |
| Parse a virtual DOM from every .html file                 | X              | X                          | X                       |                                                                                                                                                         |
| Ignore assets for build                                   | X              |                            |                         |                                                                                                                                                         |
| Rewrite AWS URLs                                          |                | X                          | X                       |                                                                                                                                                         |

> #### Publish Batch API

Author: The implementation here might shift a bit to a more generic Event Bus using SNS and SQS.  I have experience with RabbitMQ and it looks like the SNS/SQS AWS pattern most closely matches RabbitMQ functionality: https://stackoverflow.com/questions/46880229/migrate-from-amqp-to-amazon-sns-sqs-need-to-understand-concepts

Author: I've updated the architecture diagram and add a note to this section regarding SNS and SQS patter.

Editors coming from Team Sites expect publication of content within 5 minutes.  As editors migrate from Team Sites to Drupal, editors often do not understand why the content they publish does not show up on va.gov in a timely mannter.

### **High Level Design**

The same code used to generate the HTML for the preview server can be used to generate the HTML for va.gov.  

The following diagram shows a high level design of how a node in Drupal would be published “immediately” on va.gov.

Drupal is in charge of determining when and which pages need to be published based upon the action by an editor.  When Drupal determines a page is to be published, a Publish Batch Record will be sent to Amazon Simple Queue Service (SQS). A Lambda job will pick up the Publish Batch Record to build the HTML. After the HTML has been created, it’s sent to an [S3 bucket](http://content.www.va.gov.s3-website-us-gov-west-1.amazonaws.com/) which serves the html for va.gov.  

> 

Author: The Small Batch Publishing jobs are triggered from content changes in Drupal and not from code changes in a repo.  The use case for GH Actions seems to focus more on responding to events in GH.

Drupal will track the queue progress using the [advancedqueue](https://www.drupal.org/project/advancedqueue) and the Public Batch API described below. The advancedqueue module allows for observability and reporting.

#### Preview Server Changes

The code from the current preview server will be abstracted and reused in the lambda function.  The following changes will occur:

*   Support building pages for a specific node revision.  This allows the lambda function to generate HTML for a published revision or any draft revision.  
*   Add the ability to toggle [debug information](https://monosnap.com/file/e7gAeGAg7m0crd8anzjMt5Z3NiuGvL) 
*   Code will be updated to use the json processing from here: `src/site/stages/build/drupal/*.js`
*   `/diff` endpoint code will be refactored to use the same code as the lambda function.
*   GraphQL Query building will be refactored to have Full and Small Batch use the same queries.
*   The Get Drupal Content plugin should be broken apart to make the layers reusable with the different services.
*   When changes are made, validate local develpment workflow using the Preview Server with the Product and QA teams.

#### Facility Banner Alerts

> The first type of data to be updated using the Small Batch Publishing will be the facility banner alerts and alerts. These types of content currently trigger a deploy from Drupal to the jenkins job. The trigger from Drupal to Jenkins will be replaced with Small Batch Publishing.  By removing the need for Drupal to trigger the Jenkins job, we eliminate logic around Drupal bulk publish events and when an editor is allowed to trigger a Small Batch Publish.

Author: Thanks, they are two different things but we really are focused on facility banner alerts.  I'll update.

1. Banner Alert updated in Drupal
2. Drupal sends Publish Batch Record to SQS
3. Lambda picks up Publish Batch record
4. Metalsmith build runs for alerts and generates a JSON file instead of an HTML file.
5. JSON file is pushed to S3

In a sepearte effort, the code which generates the Facility Banner Alert will be moved from the Metasmith Template to a React Widget.  The React widget will load the JSON file from S3 which was generated by Drupal in the process described above.

> 

Reviewer 2: > Metasmith Template to a React Widget

Liquid, not metalsmith.

The following diagram shows a high level design of how a node in Drupal would be published “immediately” on va.gov.

Drupal is in charge of determining when and which pages need to be published based upon the action by an editor.  When Drupal determines a page is to be published, a Publish Batch Record will be sent to Amazon Simple Queue Service (SQS). A Lambda job will pick up the Publish Batch Record to build the HTML. After the HTML has been created, it’s sent to an [S3 bucket](http://content.www.va.gov.s3-website-us-gov-west-1.amazonaws.com/) which serves the html for va.gov.  

The code used by the lambda to generate the HTML will utilize the current preview server code.  As part of this work, the preview server code will be abstracted and improved to provide production ready HTML.

![image](https://lucid.app/publicSegments/view/84e99c14-7cdf-4955-95b7-badf320be509/image.png)

Editable link: [https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#](https://lucid.app/lucidchart/a8445048-94ce-4473-aafa-e8ed6cbfbad5/edit?page=0_0#)

## **Specifics**

### **Detailed Design**

> #### Drupal Triggers Content Publishing

Reviewer: are/can we address the content that lives in [vagov-content](https://github.com/department-of-veterans-affairs/vagov-content)?

Author: The content should be migrated to Drupal. I realize this is a loaded statement.

I had a brief blub in the `Mid Term Vision`

> This would require all pages in va.gov to have a Drupal representation, including the current markdown content

Reviewer: excellent, 100% agree. thanks!

> When an editor performs an action in Drupal (ie publishing a node, edits a menu item), Drupal will determine which pages need to be published. [The logic will be encapsulated by a Drupal Service.](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5509) The rules around the dependency graph are being tracked on [here](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/dependency_graph) and are currently a work in progress. Which rules are addressed will be determined by the content type being worked on to support Small Batch Publishing.  As a content type is add to Small Batch Publishing, it will be removed from the hourly content build.  Having support in only one build at a time will elimnte the risk of the full content build overriding the html from the Small Batch Publishing.

Reviewer: ```suggestion
When an editor performs an action in Drupal (ie publishing a node, edits a menu item), Drupal will determine which pages need to be published. [The logic will be encapsulated by a Drupal Service.](https://app.zenhub.com/workspaces/vagov-cms-team-5c0e7b864b5806bc2bfc2087/issues/department-of-veterans-affairs/va.gov-cms/5509) The rules around the dependency graph are being tracked on [here](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/platform/cms/accelerated_publishing/dependency_graph) and are currently a work in progress. Which rules are addressed will be determined by the content type being worked on to support Small Batch Publishing.  As a content type is add to Small Batch Publishing, it will be removed from the hourly content build.  Having support in only one build at a time will eliminate the risk of the full content build overriding the html from the Small Batch Publishing.
```

Drupal will track the queue progress using the [advancedqueue](https://www.drupal.org/project/advancedqueue) and the Public Batch API described below. The advancedqueue module allows for observability and reporting.

#### Preview Server Changes

The code from the current preview server will be abstracted and reused in the lambda function.  The following changes will occur:

*   Support building pages for a specific node revision.  This allows the lambda function to generate HTML for a published revision or any draft revision.  
*   Add the ability to toggle [debug information](https://monosnap.com/file/e7gAeGAg7m0crd8anzjMt5Z3NiuGvL) 
*   Code will be updated to use the json processing from here: `src/site/stages/build/drupal/*.js`
*   `/diff` endpoint code will be refactored to use the same code as the lambda function.
*   GraphQL Query building will be refactored to have Full and Small Batch use the same queries.
*   The Get Drupal Content plugin should be broken apart to make the layers reusable with the different services.
*   When changes are made, validate local develpment workflow using the Preview Server with the Product and QA teams.

#### Facility Banner Alerts

> The first type of data to be updated using the Small Batch Publishing will be the facility banner alerts and alerts. These types of content currently trigger a deploy from Drupal to the jenkins job. The trigger from Drupal to Jenkins will be replaced with Small Batch Publishing.  By removing the need for Drupal to trigger the Jenkins job, we eliminate logic around Drupal bulk publish events and when an editor is allowed to trigger a Small Batch Publish.

Author: ```suggestion
The first type of data to be updated using the Small Batch Publishing will be the facility banner alerts. These types of content currently trigger a deploy from Drupal to the jenkins job. The trigger from Drupal to Jenkins will be replaced with Small Batch Publishing.  By removing the need for Drupal to trigger the Jenkins job, we eliminate logic around Drupal bulk publish events and when an editor is allowed to trigger a Small Batch Publish.
```

1. Banner Alert updated in Drupal
2. Drupal sends Publish Batch Record to SQS
3. Lambda picks up Publish Batch record
4. Metalsmith build runs for alerts and generates a JSON file instead of an HTML file.
5. JSON file is pushed to S3

> In a sepearte effort, the code which generates the Facility Banner Alert will be moved from the Metasmith Template to a React Widget.  The React widget will load the JSON file from S3 which was generated by Drupal in the process described above.

Author: ```suggestion
In a separate effort, the code which generates the Facility Banner Alert will be moved from the Liquid Template to a React Widget.  The React widget will load the JSON file from S3 which was generated by Drupal in the process described above.
```
